crypto/src/asset/amount.rs

```diff
@@ -117,85 +211,110 @@ impl From<Amount> for Fq {
     }
 }
 
-impl From<u64> for Amount {
-    fn from(amount: u64) -> Amount {
+impl From<Amount> for Fr {
+    fn from(amount: Amount) -> Fr {
+        Fr::from(amount.inner)
+    }
+}
+
+impl From<u128> for Amount {
+    fn from(amount: u128) -> Amount {
         Amount { inner: amount }
     }
 }
 
-impl From<u32> for Amount {
-    fn from(amount: u32) -> Amount {
+impl From<Amount> for u128 {
+    fn from(amount: Amount) -> u128 {
+        amount.inner as u128
+    }
+}
+
+impl From<i128> for Amount {
+    fn from(amount: i128) -> Amount {
         Amount {
-            inner: amount as u64,
+            inner: amount as u128,
         }
     }
 }
 
-impl From<Amount> for u32 {
-    fn from(amount: Amount) -> u32 {
-        amount.inner as u32
+impl From<Amount> for i128 {
+    fn from(amount: Amount) -> i128 {
+        amount.inner as i128
     }
 }
 
-impl From<i64> for Amount {
-    fn from(amount: i64) -> Amount {
+impl From<[u8; 16]> for Amount {
+    fn from(bytes: [u8; 16]) -> Amount {
         Amount {
-            inner: amount as u64,
+            inner: u128::from_le_bytes(bytes),
         }
     }
 }
 
-impl From<Amount> for u128 {
-    fn from(amount: Amount) -> u128 {
-        amount.inner as u128
+impl TryFrom<Vec<u8>> for Amount {
+    type Error = anyhow::Error;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        if bytes.len() != 16 {
+            return Err(anyhow::anyhow!(
+                "could not deserialize Amount: input vec is not 16 bytes"
+            ));
+        }
+        let mut array = [0u8; 16];
+        array.copy_from_slice(&bytes);
+        Ok(Amount {
+            inner: u128::from_le_bytes(array),
+        })
+    }
+}
+
+impl Sum for Amount {
+    fn sum<I: Iterator<Item = Amount>>(iter: I) -> Amount {
+        iter.fold(Amount::zero(), |acc, x| acc + x)
     }
 }
 
-impl From<Amount> for i64 {
-    fn from(amount: Amount) -> i64 {
-        amount.inner as i64
+impl From<u64> for Amount {
+    fn from(amount: u64) -> Amount {
+        Amount {
+            inner: amount as u128,
+        }
     }
 }
 
 impl From<Amount> for u64 {
     fn from(amount: Amount) -> u64 {
-        amount.inner
+        amount.inner as u64
     }
 }
 
-impl From<Amount> for Fr {
-    fn from(amount: Amount) -> Fr {
-        Fr::from(amount.inner)
+impl TryFrom<Amount> for i64 {
+    type Error = anyhow::Error;
+    fn try_from(value: Amount) -> Result<Self, Self::Error> {
+        value
+            .inner
+            .try_into()
+            .map_err(|_| anyhow!("failed conversion!"))
     }
 }
 
 impl From<[u8; 8]> for Amount {
     fn from(bytes: [u8; 8]) -> Amount {
         Amount {
-            inner: u64::from_le_bytes(bytes),
+            inner: u64::from_le_bytes(bytes) as u128,
         }
     }
 }
```

Comment:
similarly here -- do we want to be deserializing amounts in this way, rather than as protos?